common/src/main/java/io/netty/util/concurrent/Promise.java

```diff
@@ -31,37 +29,39 @@
     /**
      * Marks this future as a success and notifies all listeners.
      *
-     * @return {@code true} if and only if successfully marked this future as a success. Otherwise {@code false} because
-     * this future is already marked as either a success or a failure.
+     * @return {@code true} if and only if successfully marked this promise as a success.
+     * Otherwise {@code false} because this promise is already marked as either a success or a failure.
      */
     boolean trySuccess(V result);
 
     /**
-     * Marks this future as a failure and notifies all listeners.
+     * Marks this promise as a failure and notifies all listeners attached to the {@linkplain #asFuture() future}.
      * <p>
      * If it is success or failed already it will throw an {@link IllegalStateException}.
      */
     Promise<V> setFailure(Throwable cause);
 
     /**
-     * Marks this future as a failure and notifies all listeners.
+     * Marks this promise as a failure and notifies all listeners.
      *
-     * @return {@code true} if and only if successfully marked this future as a failure. Otherwise {@code false} because
-     * this future is already marked as either a success or a failure.
+     * @return {@code true} if and only if successfully marked this promise as a failure.
+     * Otherwise {@code false} because this promise is already marked as either a success or a failure.
      */
     boolean tryFailure(Throwable cause);
 
     /**
-     * Make this future impossible to cancel.
+     * Make this promise impossible to cancel.
      *
-     * @return {@code true} if and only if successfully marked this future as uncancellable, or it is already done
-     * without being cancelled. Otherwise {@code false} if this future has been cancelled already.
+     * @return {@code true} if and only if successfully marked this promise as uncancellable, or it is already done
+     * without being cancelled. Otherwise {@code false} if this promise has been cancelled already.
      */
     boolean setUncancellable();
 
-    @Override
-    Promise<V> addListener(FutureListener<? super V> listener);
-
-    @Override
-    <C> Promise<V> addListener(C context, FutureContextListener<? super C, ? super V> listener);
+    /**
+     * Return the {@link Future} instance is associated with this promsie.
+     * This future will be completed upon completion of this promise.
+     *
+     * @return A future instance associated with this promise.
+     */
+    Future<V> asFuture();
```

Comment:
I would just let it stay here as otherwise if someone wants to implement their own `Promise` (for example because they want to decorate etc) it will be impossible to handle this correctly.

Reply:
I think Norman is right here. The `Promise` implementation is not necessarily based on `DefaultPromise`.